tools/src/main/java/org/apache/kafka/trogdor/agent/Agent.java

```diff
@@ -0,0 +1,337 @@
+import org.apache.kafka.trogdor.fault.FaultState;
+import org.apache.kafka.trogdor.rest.AgentFaultsResponse;
+import org.apache.kafka.trogdor.rest.CreateAgentFaultRequest;
+import org.apache.kafka.trogdor.rest.JsonRestServer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.ReentrantLock;
+
+import static net.sourceforge.argparse4j.impl.Arguments.store;
+
+/**
+ * The Trogdor agent.
+ *
+ * The agent process implements faults directly.
+ */
+public class Agent {
+    private static final Logger log = LoggerFactory.getLogger(Agent.class);
+
+    /**
+     * The clock to use for this agent.
+     */
+    private final Time time;
+
+    /**
+     * The time at which this server was started.
+     */
+    private final long startTimeMs;
+
+    /**
+     * The platform.
+     */
+    private final Platform platform;
+
+    /**
+     * The lock protecting shutdown and faultSet.
+     */
+    private final ReentrantLock lock = new ReentrantLock();
+
+    /**
+     * The condition variable which the agent thread waits on.
+     */
+    private final Condition cond = lock.newCondition();
+
+    /**
+     * The agent runnable.
+     */
+    private final AgentRunnable runnable;
+
+    /**
+     * The REST server.
+     */
+    private final JsonRestServer restServer;
+
+    /**
+     * The agent thread.
+     */
+    private final KafkaThread thread;
+
+    /**
+     * The set of pending faults.
+     */
+    private final FaultSet pendingFaults = new FaultSet();
+
+    /**
+     * The set of faults which are running.
+     */
+    private final FaultSet runningFaults = new FaultSet();
+
+    /**
+     * The set of faults which are done.
+     */
+    private final FaultSet doneFaults = new FaultSet();
+
+    /**
+     * True if the server is shutting down.
+     */
+    private boolean shutdown = false;
+
+    class AgentRunnable implements Runnable {
+        @Override
+        public void run() {
+            log.info("Starting main service thread.");
+            try {
+                while (true) {
+                    List<Fault> toStart = new ArrayList<>();
+                    List<Fault> started = new ArrayList<>();
+                    List<Fault> toEnd = new ArrayList<>();
+                    List<Fault> ended = new ArrayList<>();
+                    long now = time.milliseconds();
+                    long next_wake_ms = now + (60L * 60L * 1000L);
+                    lock.lock();
+                    try {
+                        Iterator<Fault> pending = pendingFaults.iterateByStart();
+                        while (pending.hasNext()) {
+                            Fault fault = pending.next();
+                            if (now < fault.spec().startMs()) {
```

Comment:
Yeah, it's a bit duplicative to have this logic here in the agent, as well as in the controller.  I agree that given that clock skew is a potential fault we might want to inject, we probably don't want to be relying on synchronized clocks.

I will rework this in a little bit.  In the meantime I want to post another revision that fixes some earlier comments...